File: ProblemSet_14xx/Problem_1458/solution.cs
Solution s = new();
Console.WriteLine(s.MaxDotProduct([3,-1,0],[4,5,3]));

public class Solution {
    public int MaxDotProduct(int[] nums1, int[] nums2) {
        int[] dp = new int[nums2.Length];
        for (int i = 0; i < nums1.Length; i++)
        {
            for (int j = dp.Length - 1; j >= 1 ; j--)
            {
                dp[j] = Math.Max(dp[j], dp[j - 1] + nums1[i] * nums2[j]);
            }
            dp[0] = Math.Max(dp[0], nums1[i] * nums2[0]);
            for (int j = 1; j < dp.Length; j++)
            {
                dp[j] = Math.Max(dp[j], dp[j - 1]);
            }
        }
        int max = dp[^1];
        if (max == 0)
        {
            max = nums1[0] * nums2[0];
            for (int i = 0; i < nums1.Length; i++)
            {
                for (int j = 0; j < nums2.Length; j++)
                {
                    max = Math.Max(max, nums1[i] * nums2[j]);
                }
            }
            return max;
        }
        return max;
    }
}

File: ProblemSet_14xx/Problem_1458/solution.py
from functools import cache
from typing import List


class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        @cache
        def process(index1, index2) -> int:
            processing_array = []
            if (index1 > 0):
                processing_array.append(process(index1 - 1, index2))
            if (index2 > 0):
                processing_array.append(process(index1, index2 - 1))
            if (index1 > 0 and index2 > 0):
                processing_array.append(max(process(index1 - 1, index2 - 1), 0) + nums1[index1] * nums2[index2])
            else:
                processing_array.append(nums1[index1] * nums2[index2])
            return max(processing_array)
        return process(len(nums1)-1, len(nums2)-1)
    
s = Solution()
# print("Result", s.maxDotProduct([2,1,-2,5], [3,0,-6]))
# print("Result", s.maxDotProduct([3,-2], [2,-6,7]))
print("Result", s.maxDotProduct([-1,-1], [1,1]))
print("Result", s.maxDotProduct([-3,-8,3,-10,1,3,9], [9,2,3,7,-9,1,-8,5,-1,-1]))